subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -312,6 +285,7 @@ private static void configureBuiltInTest(Project project, SourceSet mainSourceSe
             // relies on the main source set being created before the tests. So, this code here cannot live in the
             // JvmTestSuitePlugin and must live here, so that we can ensure we register this test suite after we've
             // created the main source set.
+            final SourceSet mainSourceSet = javaComponent.getSources();
```

Comment:
While it would make the `getSources()` unnecessary, a change like this would break the encapsulation that this PR is trying to establish. The `javaComponent` owns this SourceSet, so instead of fetching it "globally" through a project reference, we should instead leverage the component reference we already have to acquire the source set.

Reply:
That makes sense.  I think the concept of components owning sourcesets should be near the top of the list of things to document in the component types' javadocs.